### PR TITLE
[shardformer] Support 4D parallel + Flash Attention

### LINKED ISSUE #5762: [FEATURE]: SP with FlashAttention
### Describe the feature

Hi, I'm training LLaMA with SP and found that SP cannot be used with FlashAttention. When will SP compatible with FlashAttention?

### PR DESCRIPTION
**Pull request by Edenzzzz**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs
- [ ] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

Closes #5762

## 📝 What does this PR do?

Support 4D + Flash attn.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/booster/plugin/hybrid_parallel_plugin.py [hunk 1]
```diff
@@ -999,34 +999,37 @@ def __init__(
...
+                # assert (
+                #     tp_size > 1
+                # ), f"Sequence parallelism mode {self.sequence_parallelism_mode} must be enabled when using tensor parallelism"
+                # if sp_size != 1:
+                #     warnings.warn(
+                #         f"The sp_size will be the same as tp_size in sequence parallelism mode {self.sequence_parallelism_mode}, will ignore the given sequence parallelism size."
+                #     )
+                # self.sp_size = 1
+                self.sp_size = sp_size if sp_size is not None else 1
+                self.dp_size = dist.get_world_size() // (tp_size * pp_size * self.sp_size)
```
Comment (ver217) [inline colossalai/booster/plugin/hybrid_parallel_plugin.py:1018]: When sp mode is "split_gather" or "ring", it's actually 3d parallelism, as tp and sp share the same process group. The sp size should be always 1 here. If you want to enable "split_gather" and "ulysses" together, you should refactor linear layer.

File: colossalai/booster/plugin/hybrid_parallel_plugin.py [hunk 2]
```diff
@@ -999,34 +999,37 @@ def __init__(
...
-                self.sp_size = dist.get_world_size() if sp_size is None else sp_size
-                self.dp_size = dist.get_world_size() // (self.sp_size * pp_size)
+                # assert (
+                #     tp_size == 1
+                # ), f"Sequence parallelism mode {self.sequence_parallelism_mode} cannot be used with tensor parallelism"
+                # assert (
+                #     pp_size == 1
+                # ), f"Sequence parallelism mode {self.sequence_parallelism_mode} cannot be used with pipeline parallelism"
+                self.sp_size = dist.get_world_size() // pp_size if sp_size is None else sp_size
+                self.dp_size = dist.get_world_size() // (self.sp_size * pp_size * tp_size)
```
[the remark this answers is not in the document]
Comment (Edenzzzz) [reply colossalai/booster/plugin/hybrid_parallel_plugin.py:1027]: fixed

File: colossalai/shardformer/policies/llama.py [hunk 3]
```diff
@@ -118,24 +92,26 @@ def module_policy(self) -> Dict[Union[str, nn.Module], ModulePolicyDescription]:
...
+                "forward": get_llama_flash_attention_model_forward(
+                    self.shard_config,
+                    sp_mode=sp_mode,
+                    sp_size=sp_size,
+                    sp_group=sp_group,
+                ),
+            },
+            policy=policy,
+            target_key=LlamaModel,
+        )
```
[the remark this answers is not in the document]
Comment (Edenzzzz) [reply colossalai/shardformer/policies/llama.py]: Fixed

File: colossalai/shardformer/modeling/llama.py [hunk 4]
```diff
@@ -560,58 +616,54 @@ def forward(
...
-
         return_dict = return_dict if return_dict is not None else self.config.use_return_dict
 
         # retrieve input_ids and inputs_embeds
-        if (input_ids is None) ^ (inputs_embeds is not None):
-            raise ValueError(
-                "You cannot specify both input_ids and inputs_embeds at the same time, and must specify either one"
-            )
+        if input_ids is not None and inputs_embeds is not None:
+            raise ValueError("You cannot specify both decoder_input_ids and decoder_inputs_embeds at the same time")
```
Comment (ver217) [inline colossalai/shardformer/modeling/llama.py:623]: They are not equivalent

File: colossalai/shardformer/policies/llama.py [hunk 5]
```diff
@@ -118,16 +90,19 @@ def module_policy(self) -> Dict[Union[str, nn.Module], ModulePolicyDescription]:
             policy[attn_cls] = ModulePolicyDescription(
                 attribute_replacement=decoder_attribute_replacement,
             )
+
+        self.append_or_create_method_replacement(
```
Comment (ver217) [inline colossalai/shardformer/policies/llama.py:94]: Replace it only when sp or flash attn is enabled

File: tests/test_shardformer/test_model/test_shard_llama.py [hunk 6]
```diff
@@ -145,26 +157,39 @@ def check_forward_backward(model_fn, data_gen_fn, output_transform_fn, loss_fn,
...
+            "zero_stage": 1,
             "precision": "fp16",
             "initial_scale": 1,
         },
         {
             "tp_size": 1,
             "pp_size": 1,
             "sp_size": 2,
-            "num_microbatches": 1,
+            "num_microbatches": 2,
```
Comment (ver217) [inline tests/test_shardformer/test_model/test_shard_llama.py:177]: `num_microbatches=2` is meaningless, as pp size is 1

File: colossalai/shardformer/modeling/llama.py [hunk 7]
```diff
@@ -503,6 +519,8 @@ def forward(
         value_states = value_states.view(bsz, q_len, self.num_key_value_heads, self.head_dim).transpose(1, 2)
 
         kv_seq_len = key_states.shape[-2]
+        # if past_key_value is not None:
+        #     kv_seq_len += past_key_value[0].shape[-2]
```
[the remark this answers is not in the document]
Comment (Edenzzzz) [reply colossalai/shardformer/modeling/llama.py:523]: tuple kv cache is deprecated

File: colossalai/shardformer/policies/llama.py [hunk 8]
```diff
@@ -118,16 +90,19 @@ def module_policy(self) -> Dict[Union[str, nn.Module], ModulePolicyDescription]:
...
+        if self.shard_config.enable_flash_attention or self.shard_config.enable_sequence_parallelism:
             self.append_or_create_method_replacement(
                 description={
-                    "forward": get_llama_seq_parallel_attention_forward(sp_mode, sp_size, sp_group),
+                    "forward": get_llama_flash_attention_forward(self.shard_config, sp_mode, sp_size, sp_group),
                 },
                 policy=policy,
                 target_key=attn_cls,
             )
+        if self.pipeline_stage_manager is None:
```
Comment (ver217) [inline colossalai/shardformer/policies/llama.py:101]: You'd better add a "tab" here. LlamaModel's forward should be replaced only when sp or flash attn is enabled

**Review by ver217 — APPROVED** [review]

**Pull request merged**